[Inductor] Support vectorized transpose in CPP backend

jgong5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91532

Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed operations in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There exists one and only one fp32 load/store with outer loop var having contiguous buffer accesses. 2) When a load/store doesn't have contiguous access in an outer loop var, the access should be vectorizable from the inner-most dim. 3) No reduction. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel does the vectorized transposed load/store on tiles and then does vectorized load/store/compute along the inner-most loop axis. The vectorized transpose micro-kernel implementation borrows/refers to that from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
2. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Commit:
[Inductor] Support transpose load/store in CPP backend

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support transpose load/store in CPP backend"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support transpose load/store in CPP backend"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support transpose load/store in CPP backend"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support transpose load/store in CPP backend"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -69,3 +70,340 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
+// Code from FBGEMM:
+// https://github.com/pytorch/FBGEMM/blob/39a423e4ad1a04b77fea81c7d09c3e6f8984fae9/src/UtilsAvx512.cc#LL19C6-L19C6
+// 16 * 6 = 96 instructions
+inline void transpose_kernel_16x16_avx512(
```
It feels like this should be part of `at::vec` as opposed to in this header.

jgong5 replied:
Moved to `at::vec`.

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1049,45 +1224,187 @@ def to_dtype(x, dtype):
+class CppTile2DKernelChecker(CppVecKernelChecker):
+    """
+    We only addresss a narrow set of situations. All of the following conditions should be met:
+    1. There are only loads and stores.
```
How will this extend to kernels that do compute?  Can we do this in a more general way?

jgong5 replied:
Generalize the implementation and added compute support too.

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -902,9 +911,176 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
+
+
+@load_store_only_kernel
+class CppTile2DTailKernel(CppKernel):
```
Add docstrings

jgong5 replied:
Added.

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -902,9 +911,176 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
+
+
+@load_store_only_kernel
+class CppTile2DKernel(CppVecKernel):
```
Add docstrings

jgong5 replied:
Added.

jansel commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -69,3 +70,340 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
+    auto tmp = at::vec::Vectorized<                                                        \
+      typename std::remove_extent<typename std::remove_pointer<decltype(src)>::type>::type \
+    >::loadu(src);                                                                         \
+    tmp.store(dst);                                                                        \
```
Baking the load and store into a single macro will make this hard to generalize to kernels that have compute.

jgong5 replied:
Generalize the implementation. Now the vectorized load/store/compute reuses the logic from `CppVecKernel`.

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support compute operations in the kernel, e.g., `x.t() + x`
2. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
3. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose load/store in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed load/store in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There are only loads and stores. 2) All buffer accesses should be contiguous (stride-1) along one of the itervars. 3) There exists one and only one outer loop var having contiguous buffer accesses. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel calls the `TILE2D_COPY` and `TILE2D_COPY_TRANSPOSE` (in cpp_prefix.h) micro-kernels for vectorized memory movement. The micro-kernel implementation borrows/refers to the vectorized transposition code from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support compute operations in the kernel, e.g., `x.t() + x`
2. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
3. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented:
CI failures seem irrelevant to the PR changes.

jansel commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -69,3 +69,19 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
+
+#if defined(CPU_CAPABILITY_AVX512)
+
+#define TILE2D_COPY_TRANSPOSE(dst, src, ld_dst, ld_src)       \
```
Why do we need these macros?  Perhaps call at::vec directly?

jgong5 replied:
Thanks for the comment. I intended to let the generated code invoke something with general name. Now I reverted this and added the general `transpose_mxn` template function into at::vec. Please review. Thanks.

Commit:
Update on "[Inductor] Support vectorized transpose in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed operations in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There exists one and only one fp32 load/store with outer loop var having contiguous buffer accesses. 2) When a load/store doesn't have contiguous access in an outer loop var, the access should be vectorizable from the inner-most dim. 3) No reduction. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel does the vectorized transposed load/store on tiles and then does vectorized load/store/compute along the inner-most loop axis. The vectorized transpose micro-kernel implementation borrows/refers to that from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
2. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed operations in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There exists one and only one fp32 load/store with outer loop var having contiguous buffer accesses. 2) When a load/store doesn't have contiguous access in an outer loop var, the access should be vectorizable from the inner-most dim. 3) No reduction. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel does the vectorized transposed load/store on tiles and then does vectorized load/store/compute along the inner-most loop axis. The vectorized transpose micro-kernel implementation borrows/refers to that from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
2. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed operations in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There exists one and only one fp32 load/store with outer loop var having contiguous buffer accesses. 2) When a load/store doesn't have contiguous access in an outer loop var, the access should be vectorizable from the inner-most dim. 3) No reduction. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel does the vectorized transposed load/store on tiles and then does vectorized load/store/compute along the inner-most loop axis. The vectorized transpose micro-kernel implementation borrows/refers to that from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
2. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented:
CI failures are irrelevant.

jansel reviewed (approved):
Approving with understand this will be cleaned up in a future PR

jgong5 commented:
@pytorchbot merge

jgong5 commented:
> Approving with understand this will be cleaned up in a future PR

Thanks for all the comments/suggestions on the PR and offline. I will follow up with the tile abstraction for a cleaner design in the follow-up PRs.

Commit:
Update on "[Inductor] Support vectorized transpose in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed operations in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There exists one and only one fp32 load/store with outer loop var having contiguous buffer accesses. 2) When a load/store doesn't have contiguous access in an outer loop var, the access should be vectorizable from the inner-most dim. 3) No reduction. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel does the vectorized transposed load/store on tiles and then does vectorized load/store/compute along the inner-most loop axis. The vectorized transpose micro-kernel implementation borrows/refers to that from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
2. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed operations in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There exists one and only one fp32 load/store with outer loop var having contiguous buffer accesses. 2) When a load/store doesn't have contiguous access in an outer loop var, the access should be vectorizable from the inner-most dim. 3) No reduction. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel does the vectorized transposed load/store on tiles and then does vectorized load/store/compute along the inner-most loop axis. The vectorized transpose micro-kernel implementation borrows/refers to that from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
2. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Support vectorized transpose in CPP backend"


Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed operations in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There exists one and only one fp32 load/store with outer loop var having contiguous buffer accesses. 2) When a load/store doesn't have contiguous access in an outer loop var, the access should be vectorizable from the inner-most dim. 3) No reduction. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel does the vectorized transposed load/store on tiles and then does vectorized load/store/compute along the inner-most loop axis. The vectorized transpose micro-kernel implementation borrows/refers to that from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
2. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

jgong5 commented:
@pytorchbot merge
